R/teal_lockfile.R

```diff
@@ -0,0 +1,114 @@
+#' Generate lockfile for application reproducibility
+#'
+#' This function is invoked during [teal::init] to create `renv`-compatible lockfile for use within the application.
+#'
+#' The function leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `teal.renv.lockfile` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.renv.lockfile` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  lockfile_path <- "teal_app.lock"
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "")) {
+    if (file.exists(user_lockfile)) {
+      file.copy(user_lockfile, lockfile_path)
+      return(invisible(NULL))
+    } else {
+      stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+    }
+  }
+
+  if (!(is_in_test() || is_r_cmd_check())) {
+    old_plan <- future::plan()
```

Comment:
What could be an `old_plan`? Could it be something "planned" in a preprocessing by an app user? And by the way I love the docs of the `future::plan`.

> This function allows the user to plan the future

Reply:
So just in case some parallel process was running, we reuse this process, so that we do not close someone's parallel backend that could possibly be running some tasks.

Reply:
okey, fair enough.